### PR TITLE
ACAS-166-bb-concurrent

### PR DESCRIPTION
**Pull request by brianbolt**
Added concurrent multi-threading calls to bbchem services to speed up standardization.  I also added the ability to save and search using cmpdregmolecules instead of just the mol structure string and because I added the ability to fetch a cmpdregmolecule (using its internal representation or impl) by cd_id, I greatly improved searches and standardization by using the already saved fingerprints.  

This brought about a 5x speed improvement in the overall process

### CODE REVIEW

**Review by bffrost — COMMENTED** [review]
Left a lot of feedback, but didn't spot any clear bugs. I'm alright with you merging this, also happy to do a second round of review after you digest my feedback on this PR as well as the other.

File: src/main/java/com/labsynch/labseer/service/MetalotServiceImpl.java [hunk 1]
```diff
@@ -225,6 +225,7 @@ public MetalotReturn processAndSave(Metalot metaLot, MetalotReturn mr, ArrayList
 			String molStructure;
 			if (propertiesUtilService.getUseExternalStandardizerConfig()){
 				molStructure = chemService.standardizeStructure(parent.getMolStructure());
+				parent.setMolStructure(molStructure);
```
[the remark this answers is not in the document]
Comment (brianbolt) [reply src/main/java/com/labsynch/labseer/service/MetalotServiceImpl.java:228]: Yep. Somehow lost this during the merge of the original standardization code. :grimacing:

File: src/main/java/com/labsynch/labseer/service/StandardizationServiceImpl.java [hunk 2]
```diff
@@ -379,8 +387,12 @@ public int dupeCheckStandardizationStructures() throws CmpdRegMolFormatException
 				if (dryRunCompound.getNewMolWeight() == 0) {
 					logger.debug("mol has a weight of 0 - skipping");
 				} else {
-					hits = chemStructureService.searchMolStructures(dryRunCompound.getMolStructure(),
-							StructureType.DRY_RUN, SearchType.DUPLICATE_TAUTOMER);
+					HashMap<String, Integer> chemStructureHashMap = new HashMap<String, Integer>();
+					chemStructureHashMap.put("mol", dryRunCompound.getCdId());
+					HashMap<String, CmpdRegMolecule> cmpdRegMolecules = chemStructureService.getCmpdRegMolecules(chemStructureHashMap,
+							StructureType.DRY_RUN);
+					hits = chemStructureService.searchMolStructures(cmpdRegMolecules.get("mol"),
```
[the remark this answers is not in the document]
Comment (brianbolt) [reply src/main/java/com/labsynch/labseer/service/StandardizationServiceImpl.java:387]: chemStructureService has 5 different function signatures for searchMolStructures my `String molstructure` and I have found that to be confusing.  I actually prefer just a single function signature. In reality, I think these options should all be part of some sort "StructureSearch" class that can be setup and passed in without the need for different signatures for each use.  I'm just going to make the -1F and -1 obvious in what they are by adding some code comments.

File: src/main/java/com/labsynch/labseer/chemclasses/bbchem/BBChemStructureServiceImpl.java [hunk 3]
```diff
@@ -72,56 +80,93 @@ private String getUrlFromPreprocessorSettings(String propertyKey) throws IOExcep
...
 			logger.debug("requestString: " + requestString);
-			HttpURLConnection connection = SimpleUtil.postRequest(url, requestString, logger);
-			String postResponse = null;
-			if(connection.getResponseCode() != 200) {
-				logger.error("Error posting to fingerprint service: " + connection.getResponseMessage());
-				logger.error("Response Body: " + SimpleUtil.getStringBody(connection));
+			tasks.add(new Request(String.valueOf(i), url, requestString));
+		}
+		
+		ForkJoinPool pool = new ForkJoinPool(Runtime.getRuntime().availableProcessors());
```
[the remark this answers is not in the document]
Comment (brianbolt) [reply src/main/java/com/labsynch/labseer/chemclasses/bbchem/BBChemStructureServiceImpl.java:115]: Sounds good. Leaving this as is for now.

File: src/main/java/com/labsynch/labseer/chemclasses/bbchem/BBChemStructureServiceImpl.java [hunk 4]
```diff
@@ -162,37 +206,80 @@ private JsonNode postToProcessService(HashMap<String, String> structures)  throw
...
 
-		return responseNode;
+			// Parse the response json to get the standardized mol
+			ObjectMapper responseMapper = new ObjectMapper();
+			JsonNode responseNode = responseMapper.readTree(responseBody);
+			responseMap.put(responseId, responseNode);
+		}
+		logger.info("Got response for all " + tasks.size() + " process tasks");
+
+		// Sort the response hashmap by the keys
```
[the remark this answers is not in the document]
Comment (brianbolt) [reply src/main/java/com/labsynch/labseer/chemclasses/bbchem/BBChemStructureServiceImpl.java:268]: It's actually crucial so this is a bad bug. Working on a fix.

**Comment by brianbolt** [comment]
I requested a re-review mainly because this commit about using the request id as an int is critically import. I think I got it right this time (not sure why I chose strings as the key). Swapping and sorting by int should resolve the issue.

This is the fix commit for that issue:
https://github.com/mcneilco/acas-roo-server/pull/237/commits/0363be79a18454b400d4bc814a1a7b4eeb341178

**Review by bffrost — APPROVED** [review]
Recent changes all LGTM

**Pull request merged**